### PR TITLE
Closes #1843 resolving derive_vars_dt/derive_vars_dtm edge cases

### LINKED ISSUE #1843: Bug: `derive_vars_dt()` not imputing dates when arguments to both min_dates and max_dates are provided.
### What happened?

The following call results in dates not being imputed, which appears to be as a result of specifying both min_dates and max_dates in the same call. The same happens when imputing to "first" relative to min_dates, but also specifying max_dates. However, no error/warning is given:
derive_vars_dt(
  dataset = ex,
  dtc = EXENDTC,
  new_vars_prefix = "EXEN",
  highest_imputation = "Y",
  min_dates = exprs(EXSTDTC),
  max_dates = exprs(EOSDT),
  date_imputation = "last",
  flag_imputation = "none"
)

Resulting dataset:
image

### Session Information

_No response_

### Reproducible Example

ex <- tribble(
  ~USUBJID, ~EXTRT, ~EXDOSE, ~EXSTDTC, ~EXENDTC, ~EOSDT,
  "01-704-1233", "PLACEBO", 0, "2013-03-21", "2013-04-04", "2013-07-14",
  "01-704-1233", "PLACEBO", 0, "2013-04-05", NA, "2013-07-14",
  "01-705-1018", "PLACEBO", 0, "2013-07-05", NA, "2013-07-12",
  "01-705-1031", "XANOMELINE", 54, "2013-11-27", "2013-12-18", "2014-05-11",
  "01-705-1031", "XANOMELINE", 54, "2013-12-19", NA, "2014-05-11"
) %>% 
  mutate(EOSDT = as.Date(EOSDT), 
         EXSTDTC = as.Date(EXSTDTC))

# Non imputing (with min_dates)
example1 <- derive_vars_dt(
  dataset = ex,
  dtc = EXENDTC,
  new_vars_prefix = "EXEN",
  highest_imputation = "Y",
  min_dates = exprs(EXSTDTC),
  max_dates = exprs(EOSDT),
  date_imputation = "last",
  flag_imputation = "none"
)

# Imputing (without min_dates)
example2 <- derive_vars_dt(
  dataset = ex,
  dtc = EXENDTC,
  new_vars_prefix = "EXEN",
  highest_imputation = "Y",
  max_dates = exprs(EOSDT),
  date_imputation = "first",
  flag_imputation = "none"
)

### PR DESCRIPTION
**Pull request by zdz2101**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiraldev/devel/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/devel/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately 
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by zdz2101** [comment]
Hey @DeclanHodges can you try running your code(s) where you ran into issues using this [branch](https://github.com/pharmaverse/admiral/tree/1843_derive_vars_dt_bug%40devel)

**Comment by DeclanHodges** [comment]
> Hey @DeclanHodges can you try running your code(s) where you ran into issues using this [branch](https://github.com/pharmaverse/admiral/tree/1843_derive_vars_dt_bug%40devel) , believe I addressed that case where user supplies both min/max dates and `highest_imputation = "Y"` and `date_imputation = "last"`, added a corresponding test but certainly need an extra pair of eyes

Hey @zdz2101, I reran my code and everything runs as expected now. Thanks for looking into this!

File: R/derive_date_vars.R [hunk 1]
```diff
@@ -1016,6 +996,11 @@ convert_dtc_to_dt <- function(dtc,
     max_dates = max_dates,
     preserve = preserve
   )
+  imputed_dtc <- if_else(
+    stringr::str_starts(imputed_dtc, "(0000|9999)") | imputed_dtc %in% c("0000-01-01", "9999-12-31"), # nolint
```
Comment (bms63) [inline R/derive_date_vars.R:1000]: Do we need the `strignr::` ?

**Review by bms63 — CHANGES_REQUESTED** [review]
News entry please :)

**Review by DeclanHodges — APPROVED** [review]
Looks good to me @zdz2101!

**Review by bms63 — APPROVED** [review]
Thanks Zelos!

File: NAMESPACE [hunk 2]
```diff
@@ -270,6 +270,7 @@ importFrom(stringr,str_remove)
 importFrom(stringr,str_remove_all)
 importFrom(stringr,str_replace)
 importFrom(stringr,str_replace_all)
+importFrom(stringr,str_starts)
```
Comment (bms63) [inline NAMESPACE:273]: This is not done properly.  

```suggestion
importFrom(stringr, str_starts)
```

Comment (bms63) [reply NAMESPACE:273]: I was doing this as an example @zdz2101

**Comment by bms63** [comment]
@zdz2101 We can merge this correct?  Or we were waiting on the expressions stuff?

**Comment by zdz2101** [comment]
> @zdz2101 We can merge this correct? Or we were waiting on the expressions stuff?

I think I needed a re-approval for check templates a while back but forgot to ping, should be good to go

**Review by bms63 — APPROVED** [review]

**Pull request merged**